src/xrpld/app/tx/detail/InvariantCheck.cpp

```diff
@@ -2230,13 +2230,15 @@ ValidVault::visitEntry(
         after != nullptr && (before != nullptr || !isDelete),
         "ripple::ValidVault::visitEntry : some object is available");
 
-    // `Number balance` will capture the difference (delta) between "before"
+    // Number balanceDelta will capture the difference (delta) between "before"
     // state (zero if created) and "after" state (zero if destroyed), so the
     // invariants can validate that the change in account balances matches the
     // change in vault balances, stored to deltas_ at the end of this function.
-    Number balance{};
+    // This is done even if balanceDelta is zero, but an object was updated.
+    Number balanceDelta{};
 
-    // By default do not add anything to deltas
+    // Append to deltas if sign is non-zero, i.e. an object of an interesting
+    // type has been updated - even if the update did not change the balance.
     std::int8_t sign = 0;
```

Comment:
These comments are contradictory, and the code behavior is to only append if the sign is not zero, so the first one is wrong. If the update did not change the balance, the sign will be zero, and this the value will not be appended.

Reply:
I guess I need to change the comment to make it clear that updating an object happens whether or not its balance has changed, and that includes account root objects.

Reply:
> In the context of this block of code, `sign` will be set to -1 in `case ltACCOUNT_ROOT:` but since both `before->getFieldAmount(sfBalance)` and `after->getFieldAmount(sfBalance)` yield the same number, `balanceDelta` will be zero.

Oh, I'm foolish. I was thinking of the `sign()` of `balanceDelta`, which _is_ zero in this case, but `sign` is set manually and independently of `balanceDelta`.